pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,253 @@
+		}
+		return allErrs
+	}
+
+	if len(rule.Verbs) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("verbs"), "verbs must contain at least one value"))
+	}
+	if len(rule.APIGroups) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("apiGroups"), "resource rules must supply at least one api group"))
+	}
+	if len(rule.Resources) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("resources"), "resource rules must supply at least one resource"))
+	}
+
+	if hasWildcard(rule.Verbs) && len(rule.Verbs) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("verbs"), rule.Verbs, "if '*' is present, must not specify other verbs"))
+	}
+	if hasWildcard(rule.APIGroups) && len(rule.APIGroups) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("apiGroups"), rule.APIGroups, "if '*' is present, must not specify other api groups"))
+	}
+	if hasWildcard(rule.Resources) && len(rule.Resources) > 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("resources"), rule.Resources, "if '*' is present, must not specify other resources"))
+	}
+	if !supportedVerbs.IsSuperset(sets.NewString(rule.Verbs...)) {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("verbs"), rule.Verbs, supportedVerbs.List()))
+	}
+
+	return allErrs
+}
+
+func ValidateFlowSchemaStatus(status *flowcontrol.FlowSchemaStatus, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	// conditions will not be validated
+	return allErrs
+}
+
+func ValidatePriorityLevelConfiguration(pl *flowcontrol.PriorityLevelConfiguration) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&pl.ObjectMeta, false, ValidatePriorityLevelConfigurationName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationSpec(&pl.Spec, pl.Name, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidatePriorityLevelConfigurationStatus(&pl.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+func ValidatePriorityLevelConfigurationSpec(spec *flowcontrol.PriorityLevelConfigurationSpec, name string, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if name != flowcontrol.SystemPriorityLevelConfigurationNameTop && spec.Exempt {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("exempt"), "must not be exempt"))
+	}
+	if name != flowcontrol.SystemPriorityLevelConfigurationNameLow && spec.GlobalDefault {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("globalDefault"), "must not be global default"))
+	}
+	if spec.AssuredConcurrencyShares <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("assuredConcurrencyShares"), spec.AssuredConcurrencyShares, "must be positive"))
+	}
+	if spec.QueueLengthLimit <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("queueLengthLimit"), spec.QueueLengthLimit, "must be positive"))
+	}
+	if spec.HandSize <= 0 {
```

Comment:
`types.go` declares this field to be optional and specifies the default value, so in the case of 0 the default value should be supplied rather than a validation error produced.

Reply:
what do you think is a reasonable default value for hand-size? anyway, this comment should be addressed in defaulting func of PL objects. validation works after defaulting.

Reply:
The types doc says "If empty, the hand size will the be set to 1".  And yes, supplying the default is done earlier.  So the value 0 should never appear here, so this code is not incorrect.

Reply:
yes, then `0` is unreachable